### PR TITLE
Added Cosine Similarity loss.

### PR DESCRIPTION
**Pull request by abhinavsp0730**
- [x] Added loss

- [x]  Added test cases

- [x]  Formatted the code with black

- [x]  Added the documentation
#102
Hey, @charlielito  please review.
Thanks.

### CODE REVIEW

File: elegy/losses/cosine_similarity_test.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+    y_true = _l2_normalize(y_true, axis=1)
+    y_pred = _l2_normalize(y_pred, axis=1)
+    assert jnp.array_equal(loss, -jnp.sum(y_true * y_pred, axis=1))
+
+
+@transform_and_run
+def test_compatibility():
+    # Input:  true (y_true) and predicted (y_pred) tensors
+    y_true = jnp.array([[0.0, 1.0], [0.0, 0.0]])
```
Comment (cgarciae) [inline elegy/losses/cosine_similarity_test.py:71]: Can we use random arrays of higher dimensions here so the test is more general? We can use `np.random.uniform` to create them.

File: elegy/losses/cosine_similarity.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+         (Note dN-1 because all loss functions reduce by 1 dimension, usually axis=-1.)
+    """
+
+    def _l2_normalize(x, axis=None, epsilon=utils.EPSILON):
+        """Normalizes along dimension `axis` using an L2 norm."""
+        square_sum = jnp.sum(jnp.square(x), axis=axis, keepdims=True)
+        x_inv_norm = rsqrt(jnp.maximum(square_sum, epsilon))
+        return jnp.multiply(x, x_inv_norm)
+
+    y_true = _l2_normalize(y_true, axis=axis)
```
Comment (cgarciae) [inline elegy/losses/cosine_similarity.py:60]: This can simply be written as:

```python
y_true = y_true / jnp.linalg.norm(y_true, axis=axis, keepdims=True))
y_pred = y_pred / jnp.linalg.norm(y_pred, axis=axis, keepdims=True))

**Comment by cgarciae** [comment]
Hey @abhinavsp0730, looks amazing, thanks for adding this! :clap: 

The general structure looks really good, I left 2 comments we can sort out before merging.

File: elegy/losses/cosine_similarity.py (hunk 2 above)
Comment (abhinavsp0730) [reply elegy/losses/cosine_similarity.py:60]: Hy, @cgarciae by using this approach I'm getting sometime ```nan``` values.
for example 
```
y_true = jnp.array([[0.0, 1.0], [0.0, 0.0]]) 
y_pred = jnp.array([[0.6, 0.4], [0.4, 0.6]])
```
According to this approach if we apply CosineSimilarity in the above array we're gonna get ```nan``` values.
Till now I'm not getting what is causing the error. Kindly, help me here.

Comment (cgarciae) [reply elegy/losses/cosine_similarity.py:60]: Right! We should keep the epsilon in the denominator:

```python
y_true = y_true / jnp.maximum(jnp.linalg.norm(y_true, axis=axis, keepdims=True)), utils.EPSILON)
y_pred = y_pred / jnp.maximum(jnp.linalg.norm(y_pred, axis=axis, keepdims=True)), utils.EPSILON)
```

Comment (abhinavsp0730) [reply elegy/losses/cosine_similarity.py:60]: Hy, @cgarciae  thanks for helping me out. 
I've just one query.
```
    def _l2_normalize(x, axis=None, epsilon=utils.EPSILON):
        """Normalizes along dimension `axis` using an L2 norm."""
        square_sum = jnp.sum(jnp.square(x), axis=axis, keepdims=True)
        x_inv_norm = rsqrt(jnp.maximum(square_sum, epsilon))
        return jnp.multiply(x, x_inv_norm)
```
Assume a case where ```epsilon > square_sum```
from my implementation it would be,
``` y_true = y_true / sqrt(utils.EPSILON)```
but from your approach it'll be
``` y_true = y_true / utils.EPSILON```
Please correct me if I'm wrong.
THanks.

Comment (cgarciae) [reply elegy/losses/cosine_similarity.py:60]: This is correct. Since `EPSILON` is an arbitrary constant I don't think it doesn't matters much if its has the square root or not, if you want you can use `np.sqrt(utils.EPSILON)` instead to maintain equivalence with the TF implementation.

Comment (abhinavsp0730) [reply elegy/losses/cosine_similarity.py:60]: Thanks @cgarciae , I would like to go with ```np.sqrt(utils.EPSILON)``` . Only for the sake of maintaining TF equivalnce. Will be updating my PR with these changes very soon.

File: elegy/losses/cosine_similarity_test.py (hunk 1 above)
Comment (abhinavsp0730) [reply elegy/losses/cosine_similarity_test.py:71]: I guess there is a bug in TF implementation in ```CosineSimilarity```. I've raise an issue TF here, https://github.com/tensorflow/tensorflow/issues/44343

**Comment by abhinavsp0730** [comment]
Hy, @cgarciae thanks for helping me out with this PR. I've updated the PR with the changes that you've suggested. Kindly review again.

File: elegy/losses/cosine_similarity_test.py (hunk 1 above)
Comment (abhinavsp0730) [reply elegy/losses/cosine_similarity_test.py]: I've updated the PR kindly have a look.

**Comment by cgarciae** [comment]
Thanks @abhinavsp0730 for doing this!
